src/main/java/edu/ucla/library/iiif/fester/utils/IDUtils.java

```diff
@@ -70,58 +68,106 @@ public static String getLastPart(final String aID) {
-        final StringBuilder sb = new StringBuilder();
-        final String encodedID = URLEncoder.encode(aID, StandardCharsets.UTF_8);
+    public static String getResourceID(final String aS3Key) {
+        final String s3KeyPrefix;
+        if (aS3Key.contains(Constants.COLLECTION_S3_KEY_PREFIX)) {
+            s3KeyPrefix = Constants.COLLECTION_S3_KEY_PREFIX;
+        } else {
+            s3KeyPrefix = Constants.EMPTY;
+        }
+        return FilenameUtils.removeExtension(aS3Key.substring(s3KeyPrefix.length()));
+    }
 
-        return sb.append(aHost).append(SLASH).append(encodedID).append(SLASH).append(MANIFEST).toString();
+    /**
+     * Gets a IIIF resource ID (ARK) from its URI.
+     *
+     * @param aURI The URI of the IIIF resource
+     * @return An ID (ARK)
+     */
+    public static String getResourceID(final URI aURI) {
+        return getResourceID(getResourceS3Key(aURI));
     }
 
     /**
-     * Decodes an encoded collection ID.
+     * Gets a IIIF resource's S3 key from its URI.
      *
-     * @param aID The encoded collection ID
-     * @param aCollectionsPath The path at which collections can be found
-     * @return A decoded ID
+     * @param aURI The URI of the IIIF resource
+     * @return An S3 key
      */
-    public static String decode(final URI aID, final String aCollectionsPath) {
-        final int startIndex = aCollectionsPath.length() + 1;
+    public static String getResourceS3Key(final URI aURI) {
+        final String path = aURI.getPath();
+        final String encodedID;
+        final String uriPathPrefix;
+        final String s3KeyPrefix;
+        final int endIndex;
+
+        // Get the (encoded) resource ID from the URI path
+        if (path.contains(Constants.COLLECTION_URI_PATH_PREFIX)) {
+            uriPathPrefix = Constants.COLLECTION_URI_PATH_PREFIX;
+            s3KeyPrefix = Constants.COLLECTION_S3_KEY_PREFIX;
+            endIndex = path.length();
+        } else {
+            uriPathPrefix = Constants.SLASH;
+            s3KeyPrefix = Constants.EMPTY;
+            endIndex = path.length() - Constants.MANIFEST_URI_PATH_SUFFIX.length();
+        }
+        encodedID = path.substring(uriPathPrefix.length(), endIndex);
 
-        // Note that getPath() URL decodes the path too
-        return aID.getPath().substring(startIndex);
+        // Add any prefix and a .json extension
+        return s3KeyPrefix + URLDecoder.decode(encodedID, StandardCharsets.UTF_8) + Constants.JSON_EXT;
     }
 
     /**
-     * Decodes an encoded work ID.
+     * Gets an S3 key for a work.
      *
-     * @param aID The encoded work ID
-     * @return A decoded ID
+     * @param aID The ID (ARK) of the work
+     * @return An S3 key
      */
-    public static String decode(final URI aID) {
-        final String path = aID.getPath();
-        final int endIndex = path.lastIndexOf('/');
+    public static String getWorkS3Key(final String aID) {
+        return aID + Constants.JSON_EXT;
```

Comment:
So this is a breaking change. It won't handle requests that come in with a JSON any more. Or would you prefer to move the logic that that checks for the .json and doesn't add it if it's there here?

Reply:
I guess this raises the question too of removing the tests related to confirming that .json ended IDs works. I'm more comfortable removing them now though and treating the support for ".json" requests as deprecated. We can officially remove it from here when we go to 1.x.x.

Reply:
Is this necessary though, since requests on `/{manifestId}/manifest.json` aren't covered by Fester's HTTP API spec, and the S3 API is internal? BTW, it looks like all the S3 keys currently on the prod bucket end with `.json`, so even that shouldn't break.

However, I now realize we will need to migrate the existing collections manifests from `/collections/` to `collections/`.

Reply:
Isn't this being called from GetManifestHandler which handles Web requests (i.e. is outward facing)? There was logic there that's been removed that checked if the json extension was supplied. It didn't add the extension if it found it existed. It's not so much about what's on S3. A request with the json extension would now to go S3 with id.json.json. I realize we probably don't have any code any more that's supplying the json to the id passed to the handler (maybe?), but it is a public facing endpoint that we're changing (not in this exact location but by changing the GetManifestHandler to use this and removing the extension check from there -- line 44).

Reply:
The change could be to GetManifestHandler so it does that check there, but I thought of putting it here since you're consolidating how ID operations are handled.

Reply:
Oh I see now, the `manifestId` param (not the `manifest` path component) could have a `.json` suffix. I agree on continuing to support but deprecating.

Reply:
Yeah, maybe it just belongs in the GetManifestHandler still and keep the IDUtils clean and purposeful in what they're trying to do.